### ISSUE TITLE
FileUpload: Some files are missing when uploading multiple files

### ISSUE THREAD

**Issue by uma-ac** [laravel-filament/filament#13306]
![FileUploadMissingFile](https://github.com/filamentphp/filament/assets/121853633/f25586a3-e569-46a0-92a6-3184d6b25db1)
<img width="744" alt="FileUploadMissingFile_temp" src="https://github.com/filamentphp/filament/assets/121853633/92d08d8d-9a91-4157-b063-33c1b310f0be">
<img width="991" alt="Screenshot 2024-06-18 at 10 33 41" src="https://github.com/filamentphp/filament/assets/121853633/50f20c4d-9b1c-4b7e-95b9-c66396523b83">
### Package

filament/filament

### Package Version

3

### Laravel Version

11

### Livewire Version

V3

### PHP Version

php 8

### Problem description

When I upload multiple files, some of the files are missing.

### Expected behavior

All the files should be present in the desired folder.

### Steps to reproduce

`Section::make('')
                    ->relationship('result_files')
                    ->schema([
                    FileUpload::make('attachments')
                        ->columnSpan(12)
                        ->multiple()
                        ->maxSize(100000)
                        ->disk('uploads')
                        ->directory('result_files')
                        ->acceptedFileTypes(["application/pdf"])
                        ->getUploadedFileNameForStorageUsing(function ($file): string {
                            $fileName = str($file->getClientOriginalName());
                            $fileName_str = preg_replace('/[^a-zA-Z0-9.]/','',iconv('UTF-8', 'ASCII//TRANSLIT',  $fileName));
                            return (string) str($fileName_str)->prepend(now() . '-');
                        })
                        ->downloadable()
                        ->rules([ new MalwareFile] )
                        ->label(__('application.resource.result_documents'))
                ]),`

`    FileUpload::make('attachments')
                        ->columnSpan(12)
                        ->multiple()
                        ->maxSize(100000)
                        ->disk('uploads')
                        ->visibility('private')
                        ->directory('application_files')
                        ->getUploadedFileNameForStorageUsing(function ($file): string {
                            $fileName = str($file->getClientOriginalName());
                            $fileName_str = preg_replace('/[^a-zA-Z0-9.]/','',iconv('UTF-8', 'ASCII//TRANSLIT',  $fileName));
                            return (string) str($fileName_str)->prepend(now() . '-');
                        })
                        ->downloadable()
                        ->rules([ new MalwareFile] )
                        ->label(__('application.resource.application_documents')),`

livewire.php

`
    'temporary_file_upload' => [
        'disk' => 'tmp',        // Example: 'local', 's3'              | Default: 'default'
        "rules" => "file|mimes:pdf,jpg,jpeg,png,doc,docx,svg,mp4,mp3 | max:100000",
        // 'rules' => 'max:100000',       // Example: ['file', 'mimes:png,jpg']  | Default: ['required', 'file', 'max:12288'] (12MB)
        'directory' => null,   // Example: 'tmp'                      | Default: 'livewire-tmp'
        'middleware' => null,  // Example: 'throttle:5,1'             | Default: 'throttle:60,1'
        'preview_mimes' => [   // Supported file types for temporary pre-signed file URLs...
            'png', 'gif', 'bmp', 'svg', 'wav', 'mp4',
            'mov', 'avi', 'wmv', 'mp3', 'm4a',
            'jpg', 'jpeg', 'mpga', 'webp', 'wma', 'pdf'
        ],
        'max_upload_time' => 5, // Max duration (in minutes) before an upload is invalidated...
    ],`

### Reproduction repository (issue will be closed if this is not valid)

https://github.com/guma-005/Filament-issue.git

**Comment by zepfietje** [comment]
Please provide a non-empty reproduction repository.

**Comment by uma-ac** [comment]
updated repository

**Comment by danharrin** [comment]
Please attach the files you were uploading and explain what disk you use

**Comment by danilogiacomi** [comment]
Hi this happened to me as well, I have a different scenario and was able to find the culprit (but not a solution).

In my example I have an action opening a modal with the FileUpload in its form and the creation of one entities for each uploaded file (within the action itself).

The Resource, then, has a relationManager with the entities created in the above-mentioned action.

If in the relationManager we add a ->poll('5s') then the issue appears, if we don't it works flawlessly.

I've created a repo with the two versions of the Resources so you can try it out both the working and the non-working version.

Please find the repository at the following URL, I've added a little instruction in the README and it also is shipped with some test files to be uploaded, for your convenience.

https://github.com/danilogiacomi/filamentFileUploadProblem

**Comment by PegeDev** [comment]
any update for this issue ?

**Comment by sweebee** [comment]
Might this be related to my problem (only uploading 1 file)? https://github.com/filamentphp/filament/discussions/14446

**Comment by AndreSaSilva** [comment]
I worked around the problem by using **->maxParallelUploads(1)**
Which by default is set to 2. 

But it won't be the desired final solution.

**Comment by danilogiacomi** [comment]
> any update for this issue ?



> Might this be related to my problem (only uploading 1 file)? [#14446](https://github.com/filamentphp/filament/discussions/14446)



> I worked around the problem by using **->maxParallelUploads(1)** Which by default is set to 2.
> 
> But it won't be the desired final solution.

After commenting on this issue and seeing no movement, back then I've opened this other issue:

 https://github.com/filamentphp/filament/issues/14474 

Where I provide a repository configured to reproduce the problem.

In July the issue was closed and even though I commented that the problem is still there, the issue is still closed.

If you have the same problem, please comment on that issue and let's try to work from there...

**Comment by guma-005** [comment]
This issue is still persists. Even after Laravel 12 , Filament 3.

**Comment by danilogiacomi** [comment]
> This issue is still persists. Even after Laravel 12 , Filament 3.

Yes, please write also in the https://github.com/filamentphp/filament/issues/14474 issue, maybe they'll open it up.
In that issue I describe the problem and also provide a repository to test it out

**Comment by isaackaara** [comment]
I traced this to a race condition in the `shouldUpdateState` flag management in `file-upload.js`.

When uploading multiple files with `maxParallelUploads > 1` (default: 2), `shouldUpdateState` is set back to `true` after each individual file completes. If a Livewire poll or re-render fires while other files are still uploading, the state watcher overwrites the FilePond file list with server state, dropping in-progress uploads.

This explains both workarounds reported here:
- `->maxParallelUploads(1)` works because it serializes uploads (no race window)
- Removing `->poll()` works because it reduces the chance of a re-render hitting the vulnerability window

Fix: an `activeUploads` counter that only re-enables state updates when ALL concurrent uploads have completed.

PR: #19403

**Comment by danielh-official** [comment]
Has this bug been reproduced with later versions of FilamentPHP and Laravel?

I made one quick on my local environment to test submitting files and didn't run into any issues.

**Comment by Bramvzw** [comment]
I tried to reproduce this and want to share the results, since the April question about later versions was still open.

The poll-based path did not reproduce for me on current 4.x (v4.11.8), current 3.x (v3.3.54), or the originally reported stack (Filament 3.2.93 + Livewire 3.5.2, Laravel 11): with two parallel uploads on a throttled connection and `->poll('3s')` on the relation manager, the vulnerable window (`shouldUpdateState: true` while a file is still uploading) is open for 20+ seconds and polls tick right through it without any file loss — completed uploads leave a `livewire-file:` value in the state, which makes the `$watch('state')` handler bail before rebuilding the file list.

Two things that may explain the original reports:

1. The original report uses `getUploadedFileNameForStorageUsing()` with a `now()` prefix. Two parallel uploads finishing within the same second produce identical filenames, so one file overwrites the other — "some files are missing in the desired folder" without anything visible in the UI. That matches what @People-Sea found earlier.
2. While investigating, I did find a reproducible way to lose an in-flight upload through the same underlying race on both 3.x and 4.x: removing an already-completed file while another is still uploading. I've opened #20122 for that, with a bare reproduction repo; a fix is proposed in #20121.

**Comment by danharrin** [comment]
Given that there is now a new issue with a consistently reproducible test case, I think closing this issue is justified given the shaky reproducibility. Closing this issue doesn't stop anyone making a future PR though to improve reliability of this upload field.
